### PR TITLE
mysqlcluster,sidecar: support Clone initial when add new pod, and fix   bug after extend pvc, restore from backup fail  #250, #370

### LINKED ISSUE #250: [feature] support Clone initial when add new pod
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
When the cluster create or update rolling, create pod one by one, 
Every new Pod should clone the leader Pod and initial itself.

**Describe alternatives you've considered**
1. should backup from leader Pod, so the data is up to date.

**Additional context**

### PR DESCRIPTION
**Pull request by acekingke**
### What type of PR is this?

/enhancement
### Which issue(s) this PR fixes?

Fixes #250 

### What this PR does?

Summary:

### Special notes for your reviewer?
1. add http download endpoint, use to backup to http stream.
2. when initialize, download backup copies from http stream, and create `restore.sh` script, restore and initialize

### CODE REVIEW

**Comment by acekingke** [comment]
xenon need to be changed. 
 https://github.com/acekingke/xenon/tree/LogPosInConfig

File: sidecar/config.go [hunk 1]
```diff
@@ -323,50 +337,100 @@ func (cfg *Config) buildXenonConf() []byte {
 	}
 
 	hostName := fmt.Sprintf("%s.%s.%s", cfg.HostName, cfg.ServiceName, cfg.NameSpace)
+	var str string
+	if cfg.GtidPurged != "" {
```
Comment (runkecheng) [inline sidecar/config.go:341]: Only `gtid-purged` Different, this is a bit redundancy.

File: sidecar/config.go [hunk 2]
```diff
@@ -604,3 +668,74 @@ rm -rf /root/backup
...
+	// Xtrabackup Prepare.
+	cmd = exec.Command(xtrabackupCommand, "--defaults-file="+utils.ConfVolumeMountPath+"/my.cnf", "--use-memory=3072M", "--prepare", "--target-dir=/backup/"+cfg.XRestoreFrom)
+	cmd.Stderr = os.Stderr
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to xtrabackup prepare : %s", err)
+	}
+	// Get the backup binlong info.
+	gtid, err := GetXtrabackupGTIDPurged("/backup/" + cfg.XRestoreFrom)
+	if err == nil {
```
Comment (runkecheng) [inline sidecar/config.go:701]: Throw the error when return error?

File: sidecar/config.go [hunk 3]
```diff
@@ -604,3 +668,74 @@ rm -rf /root/backup
...
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to remove relay-bin : %s", err)
+	}
+	// Run chown -R mysql.mysql /var/lib/mysql
+	cmd = exec.Command("chown", "-R", "mysql.mysql", utils.DataVolumeMountPath)
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to chown -R mysql.mysql : %s", err)
+	}
+	log.Info("executeCloneStore sucess")
```
Comment (runkecheng) [inline sidecar/config.go:725]: fix info log

File: sidecar/config.go [hunk 4]
```diff
@@ -604,3 +668,74 @@ rm -rf /root/backup
...
+	return nil
+}
+
+func GetXtrabackupGTIDPurged(backuppath string) (string, error) {
+	b, err := ioutil.ReadFile(fmt.Sprintf("%s/xtrabackup_binlog_info", backuppath))
```
Comment (runkecheng) [inline sidecar/config.go:730]: Use variable names with actual meaning

File: sidecar/init.go [hunk 5]
```diff
@@ -44,6 +50,61 @@ func NewInitCommand(cfg *Config) *cobra.Command {
 	return cmd
 }
 
+// checke leader or service
+func CheckServiceExist(cfg *Config, service string) bool {
+	serviceURL := fmt.Sprintf("http://%s-%s:%v%s", cfg.ClusterName, service, utils.XBackupPort, "/health")
```
Comment (runkecheng) [inline sidecar/init.go:55]: We can package this to `internal` package like xenon executor.

File: sidecar/init.go [hunk 6]
```diff
@@ -44,6 +50,61 @@ func NewInitCommand(cfg *Config) *cobra.Command {
...
+
+	return true
+}
+
+// clone from leader or follower
+func runCloneAndInit(cfg *Config) error {
+	//check leader is exist?
+	serviceURL := ""
+	if len(serviceURL) == 0 && CheckServiceExist(cfg, "leader") {
+		serviceURL = fmt.Sprintf("http://%s-%s:%v", cfg.ClusterName, "leader", utils.XBackupPort)
```
Comment (runkecheng) [inline sidecar/init.go:85]: we can use `GetNameForResource`

**Review by runkecheng — CHANGES_REQUESTED** [review]

File: sidecar/init.go [hunk 7]
```diff
@@ -44,6 +50,61 @@ func NewInitCommand(cfg *Config) *cobra.Command {
...
+		// backup at first
+		Args := fmt.Sprintf("rm -rf /backup/initbackup;mkdir -p /backup/initbackup;curl --user $BACKUP_USER:$BACKUP_PASSWORD %s/download|xbstream -x -C /backup/initbackup; exit ${PIPESTATUS[0]}",
+			serviceURL)
+		cmd := exec.Command("/bin/bash", "-c", "--", Args)
+		log.Info("runCloneAndInit", "cmd", Args)
+		cmd.Stderr = os.Stderr
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("failed to disable the run restore: %s", err)
+		}
+		cfg.XRestoreFrom = "initbackup"
```
Comment (runkecheng) [inline sidecar/init.go:101]: use `const` is better?

File: sidecar/init.go [hunk 8]
```diff
@@ -44,6 +50,61 @@ func NewInitCommand(cfg *Config) *cobra.Command {
...
+		cfg.CloneFlag = true
+		return nil
+	}
+	return errors.New("not exist leader or follower")
```
Comment (runkecheng) [inline sidecar/init.go:105]: return nil and print in outer layer? This should not be error.

File: sidecar/config.go (hunk 2 above)
Comment (acekingke) [reply sidecar/config.go:701]: Cannot throw error， donot has gtid， is a normal condition.

File: sidecar/init.go (hunk 6 above)
Comment (acekingke) [reply sidecar/init.go:84]: cannot， `(c *MysqlCluster) GetNameForResource ` need define mysqlcluster, here just is sidecar

File: sidecar/init.go (hunk 5 above)
Comment (acekingke) [reply sidecar/init.go:54]: internal do not have  `timeout `

File: sidecar/server.go [hunk 9]
```diff
@@ -87,6 +103,64 @@ func (s *server) backupHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// Down hanlder.
```
Comment (andyli029) [inline sidecar/server.go:106]: Fix it.

File: sidecar/init.go [hunk 10]
```diff
@@ -44,6 +49,62 @@ func NewInitCommand(cfg *Config) *cobra.Command {
 	return cmd
 }
 
+// checke leader or service
```
Comment (andyli029) [inline sidecar/init.go:52]: Fix

File: sidecar/config.go [hunk 11]
```diff
@@ -116,6 +119,12 @@ type Config struct {
 
 	// directory in S3 bucket for cluster restore from
 	XRestoreFrom string
+
+	// Clone flag
+	CloneFlag bool
+
+	// Master logfie
```
Comment (andyli029) [inline sidecar/config.go:126]: Fix the line.

File: sidecar/config.go [hunk 12]
```diff
@@ -524,3 +540,78 @@ func (cfg *Config) executeS3Restore(path string) error {
...
+// Parse the xtrabackup_binlog_info, the format is filename \t position \t gitid \n
+// or filename \t position\n
+// Get the gtid when it is existed, or return empty string.
+// It used to purged the gtid when start the mysqld slave.
+func GetXtrabackupGTIDPurged(backuppath string) (string, error) {
+	byteStream, err := ioutil.ReadFile(fmt.Sprintf("%s/xtrabackup_binlog_info", backuppath))
+	if err != nil {
+		return "", err
+	}
+	line := strings.TrimSuffix(string(byteStream), "\n")
```
Comment (runkecheng) [inline sidecar/config.go:611]: GtidPurged may include more than one `\n`

File: sidecar/config.go (hunk 2 above)
Comment (andyli029) [reply sidecar/config.go:574]: Add the handle about the err.

File: sidecar/config.go [hunk 13]
```diff
@@ -524,3 +540,79 @@ func (cfg *Config) executeS3Restore(path string) error {
...
+	// Xtrabackup prepare and apply-log-only.
+	cmd := exec.Command(xtrabackupCommand, "--defaults-file="+utils.ConfVolumeMountPath+"/my.cnf", "--use-memory=3072M", "--prepare", "--apply-log-only", "--target-dir=/backup/"+cfg.XRestoreFrom)
+	cmd.Stderr = os.Stderr
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed to xtrabackup prepare apply-log-only : %s", err)
+	}
+	// Xtrabackup Prepare.
+	cmd = exec.Command(xtrabackupCommand, "--defaults-file="+utils.ConfVolumeMountPath+"/my.cnf", "--use-memory=3072M", "--prepare", "--target-dir=/backup/"+cfg.XRestoreFrom)
+	cmd.Stderr = os.Stderr
+	cmd.Stderr = os.Stderr
```
Comment (andyli029) [inline sidecar/config.go:568]: Delete the line.

File: sidecar/config.go (hunk 2 above)
Comment (acekingke) [reply sidecar/config.go:574]: It's normal .

**Review by runkecheng — APPROVED** [review]

**Comment by runkecheng** [comment]
lgtm

File: sidecar/server.go [hunk 14]
```diff
@@ -87,6 +103,64 @@ func (s *server) backupHandler(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// DownLoad hanlder.
```
Comment (andyli029) [inline sidecar/server.go:106]: hanlder -> handler

**Comment by andyli029** [comment]
LGTM.

**Pull request merged**